pt_miniscreen/pages/root/projects.py

```diff
@@ -64,86 +68,190 @@ def from_file(cls, file):
+            logger.info(
+                f"Project '{self.config.title}' finished with exit code {exit_code}"
+            )
+            return exit_code
+
+    def run(self):
+        logger.info(f"Starting project '{self.config.title}'")
+        user = get_user_using_first_display()
+
+        self.process = Popen(
+            split(self.config.start),
+            env=self._get_environment(),
+            cwd=self.config.path,
+            preexec_fn=lambda: switch_user(user),
+        )
+        self._handle_exit_condition()
+
+    def _handle_exit_condition(self):
+        try:
+            exit_condition = ProjectExitCondition[self.config.exit_condition.upper()]
+
+            if exit_condition == ProjectExitCondition.POWER_BUTTON_PRESS:
+                event_callback = {Message.PUB_V3_BUTTON_POWER_PRESSED: self.stop}
+            elif exit_condition == ProjectExitCondition.HOLD_X:
+                CANCEL_BUTTON_PRESS_TIME = 3
 
-    return env
+                timer = Timer(CANCEL_BUTTON_PRESS_TIME, self.stop)
 
+                def on_cancel_button_pressed():
+                    nonlocal timer
+                    timer.cancel()
+                    timer = Timer(CANCEL_BUTTON_PRESS_TIME, self.stop)
+                    timer.start()
 
-def run_project(cmd: str, cwd: str = None):
-    logger.info(f"run_project(command_str='{cmd}', cwd='{cwd}')")
-    user = get_user_using_first_display()
-    env = get_project_environment()
+                def on_cancel_button_release():
+                    nonlocal timer
+                    timer.cancel()
 
-    return Popen(
-        split(cmd),
-        env=env,
-        cwd=cwd,
-        preexec_fn=lambda: switch_user(user),
-    )
+                event_callback = {
+                    Message.PUB_V3_BUTTON_CANCEL_PRESSED: on_cancel_button_pressed,
+                    Message.PUB_V3_BUTTON_CANCEL_RELEASED: on_cancel_button_release,
+                }
+
+            self.subscribe_client = PTDMSubscribeClient()
+            self.subscribe_client.initialise(event_callback)
+            self.subscribe_client.start_listening()
+            logger.info(f"Using exit condition {exit_condition.name}")
+        except Exception:
+            logger.info("Not using an exit condition")
+
+
+class ProjectState(Enum):
+    IDLE = auto()
+    STARTING = auto()
+    RUNNING = auto()
+    STOPPING = auto()
+    ERROR = auto()
 
 
 class ProjectPage(Component):
     def __init__(self, project_config: ProjectConfig, **kwargs):
         self.project_config = project_config
-        super().__init__(**kwargs)
+        super().__init__(**kwargs, initial_state={"project_state": ProjectState.IDLE})
 
-        self.process = None
         self.text = self.create_child(
-            MarqueeText,
-            text=f"Running '{self.project_config.title}'",
-            font_size=14,
+            Text,
+            text=self.displayed_text,
+            get_text=lambda: self.displayed_text,
```

Comment:
Ah! This is why it feels a bit delayed: get_text has a default interval of 1s.

I think if you remove this line and set the text when state changes it will update immediately:

```python
    def on_state_change(self, previous_state):
        if self.state["project_state"] != previous_state["project_state"]:
            self.text.state.update({"text": self.displayed_text})
```